### DISCUSSION TITLE
Location advancement not happening when redirecting to the same URL with a different HTTP verb.

### DISCUSSION THREAD

**Discussion by panda-madness** [unpoly/unpoly#597] [Q&A] [answered]
I'm not 100% sure if this is a bug or intended behavior, so I could use some clarification.

Let's assume we have two endpoints: a `GET /companies` route that lists companies, and a `POST /companies` endpoint that creates a new company. This is a fairly common pattern.

Currently, if the `POST /companies` endpoint redirects to `GET /companies` after a successful form submission, Unpoly will not recognize that as a valid location change, and thus will not trigger all the relevant effects that should have (I stumbled upon this when I struggled with closing a modal layer upon reaching an accept location).

To work around this, one has to differentiate the URIs for these endpoints (i.e.  rename `POST /companies` to `POST /companies/store`, or add a query string parameter like `GET /companies?created=1`).

It's not a big hassle, but I would like to know if the current behavior is the intended one, since it's a bit unintuitive.

**Comment by triskweline** [comment]
> I stumbled upon this when I struggled with closing a modal layer upon reaching an accept location

Does your modal interaction also begin on `GET /companies`? Because when it begins on e.g. `GET /companies/new` then the final redirect to `GET /companies` does constitute a location change for that modal layer.

**Reply by panda-madness** [reply]
I intended to create a live reproduction demo to demonstrate this issue, but could not reproduce it. On a clean project things are indeed working as you've described here...

I will investigate further, sorry for creating a discussion before properly reproducing the error.

I will post an answer here once I figure out what exactly is wrong with my setup.

**Reply by panda-madness** [reply]
@triskweline Scratch that! I reproduced the issue that I'm facing, and hosted it [here](http://unpoly-location-demo.baimbet.com/users).

I've enabled Unpoly logging on the demo. Click the button to open a modal with a form, fill it out, then submit the form.

**Reply by triskweline** [reply, marked as answer]
I believe I see what's happening. JavaScript libraries like Unpoly never see redirects or HTTP methods. They only see the URL of the final response.

In your example Unpoly sends a request to `POST /companies` and gets a response for `/companies`. Since the URL is unchanged and since the method cannot be accessed, it assumes a response to `POST /companies` and does not advance location.

To work around this browser limitation your responses can echo the current HTTP method as an [`X-Up-Method`](https://unpoly.com/X-Up-Method) header. You may also use a library like [php-unpoly](https://github.com/webstronauts/php-unpoly) that does this for you.

> To work around this, one has to differentiate the URIs for these endpoints (i.e. rename POST /companies to POST /companies/store, or add a query string parameter like GET /companies?created=1).

If you find URL-based close conditions to be too limiting, you may also choose to [close when an event is emitted](https://unpoly.com/closing-overlays#closing-when-an-event-is-emitted). Your server can emit events using the [`X-Up-Events`](https://unpoly.com/X-Up-Events) response header, without changing your URL structure.

**Reply by panda-madness** [reply]
> To work around this browser limitation your responses can echo the current HTTP method as an [X-Up-Method](https://unpoly.com/X-Up-Method) header. You may also use a library like [php-unpoly](https://github.com/webstronauts/php-unpoly) that does this for you.

This solution works, thanks! I've created a simple middleware that appends `X-Up-Method` to all responses in my app.

> If you find URL-based close conditions to be too limiting, you may also choose to [close when an event is emitted](https://unpoly.com/closing-overlays#closing-when-an-event-is-emitted). Your server can emit events using the [X-Up-Events](https://unpoly.com/X-Up-Events) response header, without changing your URL structure.

Sadly this does not work very well for me, as custom headers are not persisted across redirects, and I'd have to write a lot more code to make my app respond directly instead of with redirects.
